builder: wire up new gc types for buildkit prune functionality

jsternberg opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

This wires up the new gc types that buildkit exposes in version 0.17. The previous flag, `KeepBytes`, was renamed to `ReservedBytes` and two new options, `MaxUsed` and `MinFree` were added.

`MaxUsed` corresponds to the maximum amount of space that buildkit will
use for the build cache and `MinFree` amount of free disk space for the
system to prevent the cache from using that space. This allows greater
configuration of the cache storage usage when used in situations where
docker is not the only service on the system using disk space.

Fixes #48639.

**- How I did it**

Modified the areas marked by @thaJeztah in https://github.com/moby/moby/issues/48639.

**- How to verify it**

Added some test cases to check that the old and new config options work.

**- Description for the changelog**

```markdown changelog
`docker buildx prune` now supports `reserved-space`, `max-used-space` and `min-free-space`, `keep-bytes` filters.
API: `POST /build/prune` renames `keep-bytes` to `reserved-space` and now supports additional prune parameters `max-used-space` and `min-free-space`.
```

**- A picture of a cute animal (not mandatory but encouraged)**

jsternberg commented on builder/builder-next/worker/gc_windows.go:
```diff
@@ -2,6 +2,8 @@
-	return defaultCap
+import "github.com/containerd/errdefs"
+
+func detectDiskSize(_ string) (int64, error) {
```
Switched to using this implementation. It also means I can just delete both of these files.

jsternberg commented on builder/builder-next/worker/gc.go:
```diff
@@ -2,52 +2,90 @@ package worker
-const defaultCap int64 = 2e9 // 2GB
+const (
+	defaultReservedSpaceBytes int64 = 2e9  // 2GB
+	defaultMaxUsedSpaceBytes  int64 = 10e9 // 10GB
```
Done.

jsternberg commented:
Updated the swagger API and found a few additional places I missed by grepping for `keep-storage`.

jsternberg commented:
Updated the default gc policy code after some offline feedback from @tonistiigi.

vvoland commented on api/server/router/build/build_routes.go:
```diff
@@ -177,19 +177,41 @@ func (br *buildRouter) postPrune(ctx context.Context, w http.ResponseWriter, r *
+		bs, err := strconv.Atoi(fv)
+		if err != nil {
+			return invalidParam{errors.Wrapf(err, "keep-storage is in bytes and expects an integer, got %v", fv)}
+		}
```
cc @thaJeztah

thaJeztah replied:
Looks like this one dropped of my radar 🙈 


Yes, I'm considering to skip this one for 27.4; IIUC, without this patch, things continue to work "as before", just the new options not being exposed (?).


For these changes above though, possibly we need some API-version gates? (need to look closer)

jsternberg commented:
This is rebased now. The failing test doesn't seem to be related.

thompson-shaun commented:
@vvoland @thaJeztah - API version check?

vvoland reviewed (changes requested):
New options should be API version gated.

`docs/api/version-history.md` also needs an entry for the newly supported options

jsternberg commented:
@vvoland done. I've never done it before so please double check that I did it correctly.

If this is API version-gated, do we need the fallback code for `keep-bytes` to keep functioning?

vvoland commented on api/swagger.yaml:
```diff
@@ -8994,14 +8994,28 @@ paths:
           type: "integer"
           format: "int64"
-        - name: "all"
+        - name: "max-used-space"
           in: "query"
-          type: "boolean"
-          description: "Remove all types of build cache"
```
I think we accidentally lost the `all` parameter 😅

vvoland commented on api/types/types.go:
```diff
@@ -169,9 +169,9 @@ type BuildCache struct {
 // BuildCachePruneOptions hold parameters to prune the build cache
 type BuildCachePruneOptions struct {
-	All         bool
-	KeepStorage int64
```
We probably should keep it as `Deprecated` for at least one version.

vvoland replied:
Bump, I think this went unaddressed :)

jsternberg replied:
Ah, I misunderstood the diff. I thought I had kept this attribute.

vvoland reviewed (commented):
Thanks, looks good but spotted some minor issues.

Could you also update the description for the release notes to be more human-facing?
Mostly thinking about showing the user how to use it. Something along the lines of "docker buildx prune now supports X, Y and Y",

jsternberg commented:
Done.

vvoland commented on api/types/types.go:
```diff
@@ -169,9 +169,11 @@ type BuildCache struct {
+	Filters       filters.Args
 
-	// FIXME(thaJeztah): add new options; see https://github.com/moby/moby/issues/48639
+	KeepStorage int64 // Deprecated: deprecated in API 1.48.
```
for a follow up, should also update: https://github.com/docker/cli/blob/master/docs/deprecated.md

vvoland commented on client/build_prune.go:
```diff
@@ -21,7 +21,16 @@ func (cli *Client) BuildCachePrune(ctx context.Context, opts types.BuildCachePru
 	if opts.All {
 		query.Set("all", "1")
 	}
-	query.Set("keep-storage", strconv.Itoa(int(opts.KeepStorage)))
```
We're still keeping `KeepStorage` as deprecated, so we should still set it.

jsternberg replied:
Restored this code.

jsternberg committed:
builder: wire up new gc types for buildkit prune functionality

This wires up the new gc types that buildkit exposes in version 0.17.
The previous flag, `KeepBytes`, was renamed to `ReservedBytes` and two
new options, `MaxUsed` and `MinFree` were added.

`MaxUsed` corresponds to the maximum amount of space that buildkit will
use for the build cache and `MinFree` amount of free disk space for the
system to prevent the cache from using that space. This allows greater
configuration of the cache storage usage when used in situations where
docker is not the only service on the system using disk space.

Signed-off-by: Jonathan A. Sternberg <jonathan.sternberg@docker.com>

vvoland reviewed (approved):
LGTM, thanks!
